Start stop for iterator

yetyetanotherusername opened the pull request:
Expand the array iteration block caching with the option to set a start and end index for the generator. This speeds up use cases where one only wants to iterate over part of an array, as previously the generator would have to iterate over the unused parts of the array as well.

This adds a new function `islice` to the array class.

TODO:
* [x] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in docs/tutorial.rst
* [x] Changes documented in docs/release.rst
* [x] AppVeyor and Travis CI passes
* [x] Test coverage is 100% (Coveralls passes)

Carreau commented:
Thanks ! It seem the email you use is not associated with your github username, I'm not sure if you care about his or not.

yetyetanotherusername commented:
Fixed, thanks!

jeromekelleher commented on zarr/core.py:
```diff
@@ -423,18 +423,31 @@ def __array__(self, *args):
         return a
 
-    def __iter__(self):
+    def islice(self, start=None, end=None):
```
There should probably be some range checking on ``start`` and ``end`` here. I guess we should follow the interface of ``itertools.islice`` as closely as possible and return the same errors they do on different out-of-range conditions.

yetyetanotherusername replied:
I played around a little with `islice`. Writing this down here so i don't forget:

- Setting a bound exceeding the array bound silently returns the array without warning or error.
- Bounds must be non negative integers (or None), otherwise a ValueError is raised.
- end > start returns [], reverse iteration is not supported at all.

I think that should cover it, or am I forgetting something?

jeromekelleher reviewed (commented):
Looks good to me. I think we just need some documentation and clarity about what happens in edge cases, following the behaviour of itertools.islice.

yetyetanotherusername commented:
I added a docstring to `islice` and added a missing edge case to the tests. Regarding documentation, since this is a minor change, are any changes in the `.rst` files needed? Does anything else from the TODO-list apply here?

Otherwise I think this is ready for review.

jeromekelleher commented on zarr/core.py:
```diff
@@ -424,6 +424,36 @@ def __array__(self, *args):
     def islice(self, start=None, end=None):
+        """
+        Yield a generator for iterating over the entire or parts of the
+        array.
```
It's probably worthwhile saying why this is useful and more efficient than iterating directly over a slice.

yetyetanotherusername commented:
Since `islice` is a public function, I also added an item to the API docs and a note to release.rst.

yetyetanotherusername commented:
Just to be sure, is there anything missing on my part?

jeromekelleher reviewed (commented):
LGTM, thanks @yetyetanotherusername!

jeromekelleher commented:
> Just to be sure, is there anything missing on my part?

It all looks good to me, I think we just need a review from one of the maintainers. @Carreau, any chance you could take a look please, or maybe recommend an appropriate reviewer?

yetyetanotherusername commented:
> LGTM, thanks @yetyetanotherusername!

Thanks for your input!

yetyetanotherusername commented:
Interesting, i can't for the life of me reproduce the mypy error that got the python3.7 pipeline to fail.

joshmoore commented:
Merged master where tests are currently green. Let's see how this goes. Thanks for your patience!

joshmoore commented:
The PR is green again! Any @zarr-developers/core-devs have thoughts?

jeromekelleher commented:
+1 from me - seems like a useful feature.

joshmoore commented:
Treating @jeromekelleher's comment as a review.

yetyetanotherusername commented:
Thanks! This closes https://github.com/zarr-developers/zarr-python/issues/615
